gems/pending/Scvmm/miq_hyperv_disk.rb

```diff
@@ -99,15 +99,28 @@ def read(size)
   def bread_cached(start_sector, number_sectors)
     $log.debug "miq_hyperv_disk.bread_cached(#{start_sector}, #{number_sectors})"
     @block_cache.keys.each do |block_range|
-      next unless block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
       sector_offset = start_sector - block_range.first
       buffer_offset = sector_offset * @block_size
-      length        = number_sectors * @block_size
-      @cache_hits[start_sector] += 1
-      return @block_cache[block_range][buffer_offset, length]
+      if block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
+        length        = number_sectors * @block_size
+        @cache_hits[start_sector] += 1
+        return @block_cache[block_range][buffer_offset, length]
+      elsif block_range.include?(start_sector)
+        sectors_in_range = block_range.last - start_sector
+        length           = sectors_in_range * @block_size
+        remaining_blocks = number_sectors - sectors_in_range
+        @cache_hits[start_sector] += 1
+        return @block_cache[block_range][buffer_offset, length] + bread_cached(block_range.last + 1, remaining_blocks)
```

Comment:
The only thing that concerns me here (and in the next if case) is that this creates a new string in memory, but doesn't cache that string, so if this same set is read over and over, we have to keep allocating.  The original code would avoid that by re-reading the whole contiguous set of blocks, but a single string was kept in the cache that didn't have to be reconstructed.